mirrord/layer/src/socket/ops.rs

```diff
@@ -1392,3 +1421,111 @@ pub(super) fn sendmsg(
 
     Detour::Success(sent_result)
 }
+
+#[cfg(target_os = "macos")]
+fn create_dns_resolver_t(
+    nameservers: impl IntoIterator<Item = socket2::SockAddr>,
+    search: impl IntoIterator<Item = CString>,
+    options: CString,
+) -> Box<dns_resolver_t> {
+    let nameserver: Vec<*mut libc::sockaddr> = nameservers
+        .into_iter()
+        .map(|sockaddr| Box::into_raw(Box::new(sockaddr.as_storage())).cast())
+        .collect();
+
+    let (nameserver, n_nameserver, _) = nameserver.into_raw_parts();
+
+    let search: Vec<*mut i8> = search
+        .into_iter()
+        .map(|sockaddr| sockaddr.into_raw().cast())
+        .collect();
+
+    let (search, n_search, _) = search.into_raw_parts();
+
+    Box::new(dns_resolver_t {
+        domain: std::ptr::null_mut(),
+        n_nameserver: n_nameserver as i32,
+        nameserver,
+        port: 0,
+        n_search: n_search as i32,
+        search,
+        n_sortaddr: 0,
+        sortaddr: std::ptr::null_mut(),
+        options: options.into_raw(),
+        timeout: 0,
+        search_order: 0,
+        if_index: 0,
+        flags: 0,
+        reach_flags: 2,
+        reserved: [0; 5],
+    })
+}
+
+#[cfg(target_os = "macos")]
+#[mirrord_layer_macro::instrument(level = "trace")]
+pub(super) unsafe fn free_dns_resolver_t(resolver: *mut dns_resolver_t) {
+    let resolver = Box::from_raw(resolver);
+
+    let nameservers = Vec::from_raw_parts(resolver.nameserver, resolver.n_nameserver as usize, 0);
+
+    for nameserver in nameservers {
+        let _ = Box::from_raw(nameserver);
+    }
+
+    let searchs = Vec::from_raw_parts(resolver.search, resolver.n_search as usize, 0);
+
+    for search in searchs {
+        let _ = CString::from_raw(search);
+    }
+
+    let _ = CString::from_raw(resolver.options);
+}
+
+#[cfg(target_os = "macos")]
+#[mirrord_layer_macro::instrument(level = "trace", ret)]
+pub(super) fn remote_dns_configuration_copy() -> Detour<*mut dns_config_t> {
+    let remote = read_remote_resolv_conf()?;
+
+    // TODO: replace error
```

Comment:
If this TODO is meant to stay there, it needs more details.

Reply:
I changed it to be more detailed